msm/src/circuit_design/witness.rs

```diff
@@ -184,8 +194,40 @@ impl<
     > LookupCap<F, CIx, LT> for WitnessBuilderEnv<F, CIx, N_WIT, N_REL, N_DSEL, N_FSEL, LT>
 {
     fn lookup(&mut self, table_id: LT, value: Vec<<Self as ColAccessCap<F, CIx>>::Variable>) {
-        let value_ix = table_id.ix_by_value(&value);
-        self.lookup_multiplicities.get_mut(&table_id).unwrap()[value_ix] += F::one();
+        let value_ix = if table_id.is_fixed() {
+            table_id
+                .ix_by_value(&value)
+                .expect("Could not resolve lookup for a fixed table")
+        } else {
+            // For runtime tables, we check if this value was already
+            // present in a table; if yes, we return its index (row),
+            // otherwise we add this value to the runtime table and
+            // return its new index (runtime_table.len() == cur_height).
+            let cur_height = self.runtime_tables.get_mut(&table_id).unwrap().len();
+            let resolver = self.runtime_tables_resolver.get_mut(&table_id).unwrap();
+            if let Some(prev_index) = resolver.get_mut(&value) {
+                *prev_index
+            } else {
+                (*resolver).insert(value.clone(), cur_height);
+                self.runtime_tables
+                    .get_mut(&table_id)
+                    .unwrap()
+                    .push(value.clone());
+                cur_height
+            }
+        };
+        {
+            let multiplicities = self.lookup_multiplicities.get_mut(&table_id).unwrap();
+            // Since we allow multiple lookups per row, runtime tables
+            // can in theory grow bigger than the domain size. We
+            // still collect multiplicities as if runtime table vector
+            // is not height-bounded, but we will split it into chunks
+            // later.
+            if !table_id.is_fixed() && value_ix > multiplicities.len() {
+                multiplicities.resize(value_ix, F::zero());
+            }
+            multiplicities[value_ix] += F::one();
```

Comment:
It's unfortunate that we're using the field here. Again, not a blocker.

Reply:
Fixed: https://github.com/o1-labs/proof-systems/pull/2432/commits/36a554ff6d396fc659df0a88b30c01d24af2fb91

I quickly checked and it seems that it only makes sense to use `u64` for multiplicities for witness building; when the multiplicities are all known and circuit is built, we have to convert them into the field, so I did not change that part since the performance would be the same it seems. Correct me if I misunderstood you.